File: play-scala-chatroom-example/app/filters/ContentSecurityPolicyFilter.scala
package filters

import javax.inject.Inject

import controllers.routes
import play.api.mvc.{EssentialAction, EssentialFilter, RequestHeader}

import scala.concurrent.ExecutionContext

/**
 * Set up a more flexible content security policy that points to self and the given
 * websocket URL.
 */
class ContentSecurityPolicyFilter @Inject()(implicit ec: ExecutionContext) extends EssentialFilter {

  override def apply(next: EssentialAction): EssentialAction = EssentialAction { request: RequestHeader =>
    val webSocketUrl = routes.HomeController.chat().webSocketURL()(request)
    next(request).map { result =>
      result.withHeaders("Content-Security-Policy" -> s"connect-src 'self' $webSocketUrl")
    }
  }
}

File: java/chatroom/app/filters/ContentSecurityPolicyFilter.java
package filters;

import controllers.routes;
import play.core.Execution;
import play.mvc.EssentialAction;
import play.mvc.EssentialFilter;
import play.mvc.Http;
import play.mvc.Result;

public class ContentSecurityPolicyFilter extends EssentialFilter {
    @Override
    public EssentialAction apply(EssentialAction next) {

        return EssentialAction.of((Http.RequestHeader requestHeader) -> {
            String webSocketUrl = routes.HomeController.chat().webSocketURL(requestHeader.asScala());
            return next.apply(requestHeader).map((Result result) ->
                    result.withHeader("Content-Security-Policy", "connect-src 'self' " + webSocketUrl), Execution.trampoline());
        });
    }
}
